### ISSUE TITLE
We can only Masks 0's in LSTM, not in any other model like CNN?

### ISSUE THREAD

**Issue by Nafees-060** [keras-team/keras#16267]
Hi, 

I want to train our model on varying input dimensions. Every input in a given batch and across batches has different dimensions.
We cannot resize our input (since we’ll lose our microscopic features). Now, since we cannot resize our input, converting them into batches of numpy array becomes impossible. In order to handle this now I have made the list for the input and each list of element contained (height, width, 1). Height is variable size and width is constant. 

Sometime my input excessively large. To handle that I am using model.fit_generator(). Here is the [link](https://github.com/himanshurawlani/fully_convolutional_network/blob/master/generator.py) from where I get the idea. In this, We find the max height and width of input in a batch and pad every other input with zeros so that every input in the batch has an equal dimension. Now we can easily convert it to a numpy array or a tensor and pass it to the fit_generator(). 

Toy Example of data after padded with 0's. Suppose I have batch of 3 samples
```
Batch1:  
sample1: 
[1 4 5 6 6
 0  0 0 0 0
 1 3 4  5 6]

sample2: 
[3 2 3 4 5
 0 0 0 0 0
 0 0 0 0 0]

sample3: 
[1 2 3 4 5
 9 2 3 4 5
 2 4 5 6 3]
```
As can be seen in the batch, one example (sample3) is quite large, so the other two (sample1 and sample2) are padded with 0 to equalize the dimensions within the batch before passing them to the model.

**Model**
Here is some line of code my model: 
```
def CNN_model(input_shape, total_classes):
    input_layer = tf.keras.Input(shape=input_shape, name="time_series_activity")
    input_mask = tf.keras.layers.Masking(mask_value=0.00000)(input_layer)
    con_l1 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(
        input_mask)  # padding="same" ensure the dimensions will be same

    con_l2 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(con_l1)

    con_l3 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(con_l2)

    con_l4 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(con_l3)
    lstm_fm = LSTM_fun(con_l4)
    output_layer = tf.keras.layers.Dense(total_classes, activation="softmax")(lstm_fm)
    return tf.keras.models.Model(inputs=input_layer, outputs=output_layer)
```

**What I am thinking:**
The model automatically **learns to ignore the zeros** (via using :  `input_mask = tf.keras.layers.Masking(mask_value=0.00000)(input_layer)`)  and learns features from the intended portion from the padded input. But It is not working. I initially set the batch size to 20 for one data set that was very small in size, but the performance was poor. This means that in a batch of 20 samples, one sample size could be `(700, 5, 1)` (maximum), and the other sample could be `(33, 5, 1)`. All samples that are smaller than the maximum size sample will be padded with 0. As the model is performing very poorly, thus, I believe the model is not ignoring the padded value. Alternatively, when I use batch size = 2, I see a great improvement in validation and testing. So it means `input_mask = tf.keras.layers. Masking(mask_value=0.00000)(input_layer)` do not work. Moreover, the training speed is also prolonged.

1- Does TensorFlow Keras provide any functionality for masking zero-padded values prior to convolution?
2- Is there a way to avoid padding the variable length input with zeros and instead feed it directly into the model? Because additional 0's have a significant impact on computations.

**System information**.
- Python version: 3.9.10
- Tensorflow version: 2.8.0-dev20211222
- Keras version: 2.6.0
- GPU Information:
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.86       Driver Version: 470.86       CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:17:00.0 Off |                  N/A |
| 48%   56C    P2   112W / 350W |  13456MiB / 24268MiB |     15%      Default |
|                               |                      |                  N/A |
+------------------------------------------------------------------------------+

**Comment by tilakrayal** [comment]
@Nafees-060 ,
 In order to expedite the trouble-shooting process, could you please provide a complete code and the TensorFlow version you are using.

**Comment by Nafees-060** [comment]
> @Nafees-060 , In order to expedite the trouble-shooting process, could you please provide a complete code and the TensorFlow version you are using.

Here is the code:

```
def LSTM_fun(cnn_featureMap):
    rl = Reshape((-1, int(cnn_SE_featureMap.shape[2]) * int(cnn_SE_featureMap.shape[3])))(
        cnn_featureMap)

    lstm_l5 = tf.keras.layers.LSTM(128, activation='tanh',
                                   recurrent_initializer=tf.keras.initializers.Orthogonal(), dropout=0.0,
                                   recurrent_dropout=0.0, return_sequences=True)(
        rl)
    lstm_l6 = tf.keras.layers.LSTM(128, activation='tanh',
                                   recurrent_initializer=tf.keras.initializers.Orthogonal(), dropout=0.9,
                                   recurrent_dropout=0.5, return_sequences=True)(lstm_l5)
    return lstm_l6

def CNN_model(input_shape, total_classes):
    input_layer = tf.keras.Input(shape=input_shape, name="time_series_activity")
    input_mask = tf.keras.layers.Masking(mask_value=0.00000)(input_layer)
    con_l1 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(
        input_mask)  # padding="same" ensure the dimensions will be same

    con_l2 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(con_l1)

    con_l3 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(con_l2)

    con_l4 = tf.keras.layers.Conv2D(64, (5, 1), activation="relu", padding="same")(con_l3)
    lstm_fm = LSTM_fun(con_l4)
    output_layer = tf.keras.layers.Dense(total_classes, activation="softmax")(lstm_fm)
    return tf.keras.models.Model(inputs=input_layer, outputs=output_layer)

 train_generator = Generator(train_x, train_y, shuffle_input_segments=True)
 test_generator = Generator(test_x, test_y)

input_shape = (None, train_x[0].shape[1], 1)
 model = CNN_model(input_shape, total_classes)
 model.summary()
 opt = tf.keras.optimizers.Adam(learning_rate=0.001)
 model.compile(loss="categorical_crossentropy",
               optimizer=opt,
               metrics=[tfa.metrics.F1Score(num_classes=total_classes, average='macro')]
               )

 path_checkpoint = os.path.join(filepath,
                                dataset,
                                "best_model_checkpoint"
                                   )
 callback_model_saving = tf.keras.callbacks.ModelCheckpoint(filepath=path_checkpoint,
                                                               monitor='val_f1_score',
                                                               mode='max',
                                                               verbose=1,
                                                               save_best_only=True)
 history = model.fit_generator(train_generator, epochs=total_epoch, steps_per_epoch=len(train_generator),
                                  callbacks=[callbacks, callback_model_saving],
                                  validation_data=test_generator, validation_steps=len(test_generator))

```

Here is the how input would be feed:
![ZYWhk](https://user-images.githubusercontent.com/63704191/159419289-63ac77eb-0212-45ce-80a6-069395640056.jpg)

I want to train our model on segment which contained varying input dimensions and one label is attach with each segment. Every input in a given batch and across batches has different dimensions. As can be seen in the image, first list element size is1095, second 16, third 66, and so on.

**Comment by tilakrayal** [comment]
@Nafees-060 ,
Code shared is full of [indentation errors](https://colab.research.google.com/gist/tilakrayal/5155366a2eccdf0818cc8b49b9caa96f/untitled262.ipynb), please share a colab gist with issue reported or complete indented code with all dependencies such that we can replicate the issue reported.

**Comment by Nafees-060** [comment]
@tilakrayal  I've fixed the indentation errors. Kindly [verify](https://colab.research.google.com/gist/Nafees-060/a2807e4f24b77d84b745ac3359d38504/untitled262.ipynb#scrollTo=jnhiCVCpHlzK). Apart from that, I've provided a link to the Generator class, which you may attempt to use. If not, then simply attempt to generate random data that is consistent with the nature of my data, as I stated in the post and in the image. Then you can feed directly into the model.

**Comment by gadagashwini** [comment]
@Nafees-060, 
Padding with zero's doesn't affect the overall accuracy. When we do padding, length of the longest sequence will be considered, so even for that batch all neurons will derive some activation, the network will definitely not learn to predict just sequences of 0s because you have not provided them ever. 

Configure a [keras.layers.Embedding](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding) layer with mask_zero=True
